File: typescript-backend/lib/engines/MetricsAgent/MetricsAgent.ts
import { AgentBase } from "./AgentBase";
import { eventBus } from "../core/engineManager";

/**
 * MetricsAgent
 * ------------
 * Collects, processes, and stores system and application metrics.
 */
export class MetricsAgent extends AgentBase {
    private metricsStore: Record<string, number> = {};

    constructor() {
        super("MetricsAgent");
        this.subscribeToDBEvents();
    }

    /**
     * Subscribe to database or system events
     */
    private subscribeToDBEvents() {
        eventBus.subscribe("db:update", (event) => this.handleDBUpdate(event));
        eventBus.subscribe("db:delete", (event) => this.handleDBDelete(event));
    }

    /**
     * Handle database updates
     */
    async handleDBUpdate(event: any) {
        const { collection, key, value } = event;
        if (collection !== "metrics") return;
        this.metricsStore[key] = value;
        console.log(`[MetricsAgent] Metric updated: ${key}`, value);
    }

    /**
     * Handle database deletion
     */
    async handleDBDelete(event: any) {
        const { collection, key } = event;
        if (collection !== "metrics") return;
        delete this.metricsStore[key];
        console.log(`[MetricsAgent] Metric deleted: ${key}`);
    }

    /**
     * Set a metric
     */
    async setMetric(key: string, value: number) {
        this.metricsStore[key] = value;
        eventBus.publish("db:update", { collection: "metrics", key, value });
        console.log(`[MetricsAgent] Metric saved: ${key}`);
        return { success: true };
    }

    /**
     * Get a metric
     */
    async getMetric(key: string) {
        return this.metricsStore[key] || 0;
    }

    /**
     * Remove a metric
     */
    async removeMetric(key: string) {
        delete this.metricsStore[key];
        eventBus.publish("db:delete", { collection: "metrics", key });
        console.log(`[MetricsAgent] Metric removed: ${key}`);
        return { success: true };
    }

    /**
     * Default run method
     */
    async run(input: any) {
        console.log(`[MetricsAgent] Run called with input:`, input);
        return { success: true };
    }

    /**
     * Recover from errors
     */
    async recover(err: any) {
        console.warn(`[MetricsAgent] Recovering from error`, err);
    }
}

File: python-backend/backend-python/agents/MetricsAgent.py
# backend-python/agents/MetricsAgent.py

from core.agent_base import AgentBase
from core.db_manager import db
from core.event_bus import event_bus
from utils.logger import logger

class MetricsAgent(AgentBase):
    name = "MetricsAgent"

    def __init__(self):
        super().__init__()

    async def record_metric(self, metric_name: str, value: float):
        """
        Records a single metric to the DB and emits a db:update event.
        """
        metric_data = {
            "name": metric_name,
            "value": value
        }

        # Save to DB
        await db.set("metrics", metric_name, metric_data, "edge")
        event_bus.publish("db:update", {
            "collection": "metrics",
            "key": metric_name,
            "value": metric_data,
            "source": self.name
        })

        logger.log(f"[{self.name}] Metric recorded: {metric_name} = {value}")
        return metric_data

    async def recover(self, error: Exception):
        logger.error(f"[{self.name}] MetricsAgent error: {error}")
